modules/kargoBidAdapter.js

```diff
@@ -4,286 +4,484 @@ import { registerBidder } from '../src/adapters/bidderFactory.js';
+  }
+
+  if (bid.bidRequestsCount > 0) {
+    imp.bidRequestCount = bid.bidRequestsCount
+  }
+
+  if (bid.bidderRequestsCount > 0) {
+    imp.bidderRequestCount = bid.bidderRequestsCount
+  }
+
+  if (bid.bidderWinsCount > 0) {
+    imp.bidderWinCount = bid.bidderWinsCount
+  }
+
+  const gpid = getGPID(bid)
+  if (gpid != null && gpid != '') {
+    imp.fpd = {
+      gpid: gpid
+    }
+  }
+
+  if(bid.mediaTypes != null) {
+    if (bid.mediaTypes.banner != null) {
+      imp.banner = bid.mediaTypes.banner
     }
-    return userIds;
-  },
-
-  _getClientId() {
-    const crb = spec._getCrb();
-    return crb.clientId;
-  },
-
-  _getAllMetadata(bidderRequest, tdid) {
-    return {
-      userIDs: spec._getUserIds(tdid, bidderRequest.uspConsent, bidderRequest.gdprConsent),
-      pageURL: bidderRequest?.refererInfo?.page,
-      rawCRB: storage.getCookie('krg_crb'),
-      rawCRBLocalStorage: spec._getLocalStorageSafely('krg_crb')
-    };
-  },
 
-  _getSessionId() {
-    if (!sessionId) {
-      sessionId = spec._generateRandomUuid();
+    if (bid.mediaTypes.video != null) {
+      imp.video = bid.mediaTypes.video
     }
-    return sessionId;
-  },
 
-  _getRequestCount() {
-    if (lastPageUrl === window.location.pathname) {
-      return ++requestCounter;
+    if (bid.mediaTypes.native != null) {
+      imp.native = bid.mediaTypes.native
     }
-    lastPageUrl = window.location.pathname;
-    return requestCounter = 0;
-  },
-
-  _generateRandomUuid() {
-    try {
-      // crypto.getRandomValues is supported everywhere but Opera Mini for years
-      var buffer = new Uint8Array(16);
-      crypto.getRandomValues(buffer);
-      buffer[6] = (buffer[6] & ~176) | 64;
-      buffer[8] = (buffer[8] & ~64) | 128;
-      var hex = Array.prototype.map.call(new Uint8Array(buffer), function(x) {
-        return ('00' + x.toString(16)).slice(-2);
-      }).join('');
-      return hex.slice(0, 8) + '-' + hex.slice(8, 12) + '-' + hex.slice(12, 16) + '-' + hex.slice(16, 20) + '-' + hex.slice(20);
-    } catch (e) {
-      return '';
+  }
+
+  return imp
+}
+
+function getGPID(bid) {
+  if (bid.ortb2Imp != null) {
+    if (bid.ortb2Imp.gpid != null & bid.ortb2Imp.gpid != '') {
+      return bid.ortb2Imp.gpid
     }
-  },
 
-  _sendTimeoutData(auctionId, auctionTimeout) {
-    let params = {
-      aid: auctionId,
-      ato: auctionTimeout,
-    };
+    if (bid.ortb2Imp.ext != null && bid.ortb2Imp.ext.data != null) {
+      if (bid.ortb2Imp.ext.data.pbAdSlot != null && bid.ortb2Imp.ext.data.pbAdSlot != '') {
+        return bid.ortb2Imp.ext.data.pbAdSlot
+      }
 
-    try {
-      let timeoutRequestUrl = buildUrl({
-        protocol: 'https',
-        hostname: 'krk.kargo.com',
-        pathname: '/api/v1/event/timeout',
-        search: params
-      });
+      if (bid.ortb2Imp.ext.data.adServer.adSlot != null && bid.ortb2Imp.ext.data.adServer.adSlot != '') {
+        return bid.ortb2Imp.ext.data.adServer.adSlot
+      }
+    }
+  }
```

Comment:
just trying to understand some of the things here, thought I'd ask if it's imperative to check if `adServer` is null before proceeding to adSlot?

Reply:
Yes you are correct. I missed that one. I'll update